src/main/java/com/ebay/myriad/scheduler/NMHeartBeatHandler.java

```diff
@@ -0,0 +1,154 @@
+        yarnNodeCapacityMgr.removeCapacity(rmNode);
+      }
+      break;
+
+      case EXPIRE: {
+        nodeStore.remove(event.getNodeId().getHost());
+      }
+      break;
+
+      case STATUS_UPDATE: {
+        handleStatusUpdate(event, context);
+      }
+      break;
+
+      default: {
+        LOGGER.debug("Unhandled event: {}", event.getClass().getName());
+      }
+      break;
+    }
+  }
+
+  private void handleStatusUpdate(RMNodeEvent event, RMContext context) {
+    if (!(event instanceof RMNodeStatusEvent)) {
+      LOGGER.error("{} not an instance of {}", event.getClass().getName(),
+          RMNodeStatusEvent.class.getName());
+      return;
+    }
+
+    RMNodeStatusEvent statusEvent = (RMNodeStatusEvent) event;
+    RMNode rmNode = context.getRMNodes().get(event.getNodeId());
+    SchedulerNode schedulerNode = yarnScheduler.getSchedulerNode(rmNode.getNodeID());
+    String hostName = rmNode.getNodeID().getHost();
+
+    int rmSchNodeNumContainers = schedulerNode.getNumContainers();
+    List<ContainerStatus> nmContainers = statusEvent.getContainers();
+
+    if (nmContainers.size() != rmSchNodeNumContainers) {
+      LOGGER.warn("Node: {}, Num Containers known by RM scheduler: {}, "
+          + "Num containers NM is reporting: {}",
+          hostName, rmSchNodeNumContainers, nmContainers.size());
+    }
+
+    // Send task update to Mesos
+    for (ContainerStatus status : nmContainers) {
+      ContainerId containerId = status.getContainerId();
+      Protos.SlaveID slaveId = nodeStore.getNode(hostName).getSlaveId();
+      if (status.getState() == ContainerState.COMPLETE) {
+        LOGGER.debug("Task complete: {}", containerId);
+        yarnNodeCapacityMgr.removeCapacity(rmNode, containerId);
+        requestExecutorToSendTaskStatusUpdate(
+            slaveId,
+            containerId, Protos.TaskState.TASK_FINISHED);
+      } else { // state == NEW | RUNNING
+        requestExecutorToSendTaskStatusUpdate(
+            slaveId,
+            containerId, Protos.TaskState.TASK_RUNNING);
+      }
+    }
+  }
+
+  /**
+   * sends a request to executor on the given slave to send back a status update
+   * for the mesos task launched for this container.
+   *
+   * @param slaveId
+   * @param containerId
+   * @param taskState
+   */
+  private void requestExecutorToSendTaskStatusUpdate(Protos.SlaveID slaveId,
+      ContainerId containerId,
+      Protos.TaskState taskState) {
+    ContainerTaskStatusRequest containerTaskStatusRequest = new ContainerTaskStatusRequest();
+    containerTaskStatusRequest.setMesosTaskId(
+        ContainerTaskStatusRequest.YARN_CONTAINER_TASK_ID_PREFIX + containerId.toString());
+    containerTaskStatusRequest.setState(taskState.name());
+    myriadDriver.getDriver().sendFrameworkMessage(
+        getExecutorId(slaveId),
+        slaveId,
+        new Gson().toJson(containerTaskStatusRequest).getBytes(Charset.defaultCharset()));
+  }
```

Comment:
@rkannan82 I think @DarinJ is saying that since framework message (to executor) is unreliable, there is a potential problem here in that the executor may not receive the framework message and hence may not send out a task COMPLETED status update to mesos preventing mesos from reclaiming resources from Myriad. 

@DarinJ - the original intent was not to implement it this way. But I ended up doing it this way since myriad executor doesn't know the statuses of each YARN container. I think what we need to do is as follows:

1. Instead of having myriad executor separate from NM, we combine them into a single process that has both NM and the executor. 
2. First step toward merging executor and NM is to achieve the current executor's functionality in a different way. Today myriad executor prepares the appropriate configuration (cpu/mem capacities, cgroups hierarchy for YARN, dynamic ports etc) before launching the Node Manager. The equivalent functionality can be achieved by a java class that wraps around YARN's NodeManager. The wrapper simply prepares the configuration (the way executor is doing today), and instantiates NodeManager class with this configuration. (@DarinJ had exactly the same idea a while ago, I don't remember the context though.)
3. We still need an executor that is capable of sending out status updates for the mesos tasks (pseudo mesos tasks) launched for each YARN container. For this, we can implement an executor as NM's auxiliary service. Auxiliary service is simply a pluggable java class that can be added to NodeManager using YARN's configuration. The MapReduce http shuffle end point is implemented this way. The AuxiliaryService interface provides "listener" methods that are invoked by NodeManager whenever YARN containers are launched/finished. By implementing mesos executor under the AuxiliaryService interface, the executor becomes capable of sending out the appropriate mesos task statuses as and when YARN containers are launched/finished.

Thoughts?